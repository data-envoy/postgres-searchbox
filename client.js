export function make_client(searchApiUrl) {
  return {
    search: async (queries) => {
      const resp = await fetch(searchApiUrl, { method: 'POST', body: JSON.stringify(queries[0]) })
      if (!resp.ok) { throw new Error(resp.body) }
      return await resp.json()
    },
  }
}
