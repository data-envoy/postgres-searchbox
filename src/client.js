export function make_client(searchApiUrl) {
  return {
    search: async (queries) => {
      const resp = await fetch(searchApiUrl, {
        method: 'POST',
        body: JSON.stringify(queries[0]),
        headers: { 'Content-Type': 'application/json' },
      });
      if (!resp.ok) {
        throw new Error(resp.body);
      }
      return await resp.json();
    },
  };
}
