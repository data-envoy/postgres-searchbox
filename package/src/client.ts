import { Request } from './client.types';

// export interface ClientOptions {
//   highlightColumns?: string[];
//   returnColumns?: string[];
//   language?: string;
// }

export function make_client(
  searchApiUrl: string
  // , pgOptions?: ClientOptions
) {
  return {
    search: async (requests: Request[]) => {
      const resp = await fetch(searchApiUrl, {
        method: 'POST',
        body: JSON.stringify({
          requests,
          // pgOptions
        }),
        headers: { 'Content-Type': 'application/json' },
      });
      if (!resp.ok) {
        throw new Error(resp.body?.toString());
      }
      return await resp.json();
    },
  };
}
