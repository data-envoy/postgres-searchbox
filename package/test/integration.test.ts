import express from 'express';
import type { Server } from 'http';
import pkg from 'pg';
const { Client } = pkg;
import format from 'pg-format';
// Scripts
import { initTestDatabase } from '@scripts/mock-data.js';
import { createColumnAndIndex } from '@scripts/create-index.js';
// Main functions
import { make_client } from '@/client.js';
import { getSearchHandler } from '@/index.js';

/**
 * This file is for testing the integration of server with client.
 * Although /scripts are imported here thay are not being tested
 * they're used to set the database state for the tests
 */

describe('integration', () => {
  const tableName = 'test_table';
  const initTestDatabaseParams = { tableName, rowCount: 100, fakerSeed: 123 };

  const client = new Client();
  client.connect();

  let serverListener: Server;

  beforeAll(async () => {
    /**
     * Start an express server with searchHandler on the route /api/search
     */

    const app = express();
    const port = 3002;
    app.use(express.json());

    app.post('/api/search', getSearchHandler());
    app.post(
      '/api/search-with-options',
      getSearchHandler([
        {
          tableName,
          validReturnColumns: ['id', 'name', 'description', 'price'],
          validHighlightColumns: ['name', 'description'],
        },
      ])
    );

    serverListener = app.listen(port);
  });

  beforeEach(async () => {
    // Drop the table so that the tests can be run independently
    const dropSql = format('DROP TABLE IF EXISTS %I', tableName);
    await client.query(dropSql);
  });

  afterAll(async () => {
    // Tairdown express
    await serverListener.close();
    // Drop and close the database connection
    const dropSql = format('DROP TABLE IF EXISTS %I', tableName);
    await client.query(dropSql);
    await client.end();
  });

  it('should return results from real server: no options', async () => {
    // Prerequisites
    await initTestDatabase(initTestDatabaseParams);
    await createColumnAndIndex({ tableName });

    const client = make_client('http://localhost:3002/api/search');

    const response = await client.search([
      {
        indexName: tableName,
        params: { query: 'affordable keyboard' },
      },
    ]);

    expect(response.results[0].hits[0]).toMatchSnapshot();
  });

  it('should return results from real server: with options', async () => {
    // Prerequisites
    await initTestDatabase(initTestDatabaseParams);
    await createColumnAndIndex({ tableName });

    const client = make_client(
      'http://localhost:3002/api/search-with-options',
      {
        returnColumns: ['id', 'name', 'description', 'price'],
        highlightColumns: ['name', 'description'],
      }
    );

    const response = await client.search([
      {
        indexName: tableName,
        params: {
          query: 'affordable keyboard',
          highlightPreTag: '__ais-highlight__',
          highlightPostTag: '__/ais-highlight__',
        },
      },
    ]);

    expect(response.results[0].hits[0]).toMatchSnapshot();
  });
});
