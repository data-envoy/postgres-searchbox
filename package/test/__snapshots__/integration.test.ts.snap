// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration should return results from real server: no options 1`] = `
{
  "brand": "Powlowski, Brekke and Hamill",
  "description": "The slim & simple Maple Gaming Keyboard from Dev Byte comes with a sleek body and 7- Color RGB LED Back-lighting for smart functionality",
  "id": 20,
  "name": "Bespoke Granite Keyboard",
  "price": 14,
}
`;

exports[`integration should return results from real server: with options 1`] = `
{
  "_highlightResult": {
    "description": {
      "fullyHighlighted": false,
      "matchLevel": "full",
      "matchedWords": [],
      "value": "The slim & simple Maple Gaming __ais-highlight__Keyboard__/ais-highlight__ from Dev Byte comes with a sleek body and",
    },
    "name": {
      "fullyHighlighted": false,
      "matchLevel": "full",
      "matchedWords": [],
      "value": "__ais-highlight__Bespoke__/ais-highlight__ Granite __ais-highlight__Keyboard__/ais-highlight__",
    },
  },
  "description": "The slim & simple Maple Gaming Keyboard from Dev Byte comes with a sleek body and 7- Color RGB LED Back-lighting for smart functionality",
  "id": 20,
  "name": "Bespoke Granite Keyboard",
  "price": 14,
}
`;
